[clang][OpenMP] 6.0: Add default clause support for 'target' directive

ddpagan opened the pull request:
Per OpenMP 6.0 specification, section 7.5.1, default Clause

Page 224, lines 3-5 default Clause, Semantics
  If data-sharing-attribute is shared then the clause has no effect
  on a target construct; otherwise, its effect on a target construct is
  equivalent to specifying the defaultmap clause with the same
  data-sharing-attribute and variable-category.

Testing:
  OpenMP LIT tests
  check-all

ddpagan committed:
[clang][OpenMP] 6.0: Add default clause support for 'target' directive

Per OpenMP 6.0 specification, section 7.5.1, default Clause

Page 224, lines 3-5 default Clause, Semantics
  If data-sharing-attribute is shared then the clause has no effect
  on a target construct; otherwise, its effect on a target construct is
  equivalent to specifying the defaultmap clause with the same
  data-sharing-attribute and variable-category.

Testing:
  OpenMP LIT tests
  check-all

ddpagan committed:
Updated OpenMPSupport.rst file

ddpagan committed:
Per comment:
  Restructured/merged code that selects 'defaultmap' behavior or normal
  data-sharing-attributes depending on if associated directive is
  'target'.

alexey-bataev commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -17316,45 +17316,101 @@ OMPClause *SemaOpenMP::ActOnOpenMPDefaultClause(
+    return nullptr;
+  }
+  bool IsTargetDefault = getLangOpts().OpenMP >= 60 &&
+                         DSAStack->getCurrentDirective() == OMPD_target;
```
What about combined directives?

ddpagan committed:
Per comments:
  Updated check for target directive to allow for combined target
  directives.

  Moved code that determines data-sharing attributes into lambda that
  sets appropriate data-sharing attributes for non-target directives
  and defaultmap for target directives.

  Removed uses of std::function.

  Added test for combined target directive.

alexey-bataev reviewed (approved):
LG with a nit

ddpagan committed:
Modified lambda definition to use "auto" instead of "auto &&".

ddpagan commented on clang/lib/Sema/SemaOpenMP.cpp:
```diff
@@ -17316,45 +17316,102 @@ OMPClause *SemaOpenMP::ActOnOpenMPDefaultClause(
+  if (IsTargetDefault && M == OMP_DEFAULT_shared)
+    return nullptr;
+
+  auto &&SetDefaultClauseAttrs = [&](llvm::omp::DefaultKind M,
```
Done. Thanks, Alexey.
